src/otx/core/engine/train.py

```diff
@@ -0,0 +1,88 @@
+"""Engine component to training pipeline."""
+from __future__ import annotations
+
+import logging as log
+from typing import TYPE_CHECKING, Any
+
+import hydra
+
+from otx.core.config import TrainConfig
+
+if TYPE_CHECKING:
+    from lightning import Callback, LightningModule, Trainer
+    from lightning.pytorch.loggers import Logger
+
+
+def train(cfg: TrainConfig) -> tuple[dict[str, Any], dict[str, Any]]:
+    """Trains the model. Can additionally evaluate on a testset, using best weights obtained during training.
+
+    This method is wrapped in optional @task_wrapper decorator, that controls the behavior during
+    failure. Useful for multiruns, saving info about the crash, etc.
+
+    :param cfg: A DictConfig configuration composed by Hydra.
+    :return: A tuple with metrics and dict with all instantiated objects.
```

Comment:
I think it'd be good to turn this into a method in `Engine`, so it could be used as an API instead of CLI.

Reply:
I'm curious why do you stick to class based design this time contrary to https://github.com/openvinotoolkit/training_extensions/pull/2670#discussion_r1403620792.

From your perspective, it could be simplified to a function. The same functionality could be achieved with a function definition.

Reply:
`OTXDatasetFactory` currently lacks an __init__ method or any instance methods, and only contains a class method. Therefore, in my opinion, it could be simplified to a function. I suggest converting `train` into a method within `Engine` class for clarity, as it would be more intuitive to have an Engine with train, test, export, and optimize methods.

The concept of Engine would be similar to what Harim created in otx/v2. However, this time we would utilize Hydra for configuration and PyTorch Lightning for the model. Essentially, Hydra initialization would occur during Engine.__init__. A conceptual draft could look something like this:

```python
class Engine:
    def __init__(self, cfg):
        model: LightningModule = hydra.utils.instantiate(cfg.model)

        log.info("Instantiating callbacks...")
        callbacks: list[Callback] = instantiate_callbacks(cfg.callbacks)

        log.info("Instantiating loggers...")
        logger: list[Logger] = instantiate_loggers(cfg.logger)

        self.trainer: Trainer = hydra.utils.instantiate(
        cfg.trainer,
        callbacks=callbacks,
        logger=logger,
    )

    def train(self):
        # necessary steps can be added below
        # ...
        return self.trainer.train()
    
    def predict(self, image):
        ...

    def export(self, model, output_path):
        return export(
            self.cfg.deploy_cfg,
            model,
            f"{output_path}/openvino",
            half_precision,
            onnx_only=export_format == ExportType.ONNX,
        )
```

Reply:
https://github.com/openvinotoolkit/training_extensions/pull/2670/commits/000ecdd3194bd6898718c00b87d99128b1c8edd5 added a skeleton for this.